packages/aws-cdk-lib/aws-batch/lib/managed-compute-environment.ts

```diff
@@ -1131,15 +1141,61 @@ export class FargateComputeEnvironment extends ManagedComputeEnvironmentBase imp
   }
 }
 
-function renderInstances(types?: ec2.InstanceType[], classes?: ec2.InstanceClass[], useOptimalInstanceClasses?: boolean): string[] {
-  const instances = [];
+function checkArchitectureConsistency(
+  newInstanceType: ec2.InstanceType,
+  existingTypes: ec2.InstanceType[],
+  existingClasses: ec2.InstanceClass[],
+): void {
+  const newArchitecture = newInstanceType.architecture;
+
+  for (const existingType of existingTypes ?? []) {
+    if (existingType.architecture !== newArchitecture) {
+      throw new Error(`Cannot add instance type ${newInstanceType.toString()} with architecture ${newArchitecture}. It conflicts with existing instance type ${existingType.toString()} with architecture ${existingType.architecture}.`);
+    }
+  }
+
+  for (const existingClass of existingClasses ?? []) {
+    const tempInstanceType = new ec2.InstanceType(`${existingClass.toString()}.large`);
+    const existingClassArchitecture = tempInstanceType.architecture;
+    if (existingClassArchitecture !== newArchitecture) {
+      throw new Error(`Cannot add instance type ${newInstanceType.toString()} with architecture ${newArchitecture}. It conflicts with existing instance class ${existingClass} with architecture ${existingClassArchitecture}.`);
+    }
+  }
+}
 
+function renderInstances(
+  scope: Construct, types?: ec2.InstanceType[], classes?: ec2.InstanceClass[], useOptimalInstanceClasses?: boolean): string[] {
+  const instances = [];
+  let hasArmInstances = false;
+  let hasX86Instances = false;
   for (const instanceType of types ?? []) {
     instances.push(instanceType.toString());
+    if (instanceType.architecture === ec2.InstanceArchitecture.ARM_64) {
+      hasArmInstances = true;
+    } else {
+      hasX86Instances = true;
+    }
   }
+
   for (const instanceClass of classes ?? []) {
     instances.push(instanceClass);
+    const tempInstanceType = new ec2.InstanceType(`${instanceClass.toString()}.large`);
+    if (tempInstanceType.architecture === ec2.InstanceArchitecture.ARM_64) {
+      hasArmInstances = true;
+    } else {
+      hasX86Instances = true;
+    }
   }
+
+  if (hasArmInstances && hasX86Instances) {
+    throw new Error('Cannot mix ARM and x86 instance types or classes');
+  }
+
+  if (hasArmInstances && (useOptimalInstanceClasses || useOptimalInstanceClasses === undefined)) {
+    Annotations.of(scope).addWarningV2('@aws-cdk/aws-batch:optimalNotSupportedWithARM', '\'optimal\' instance types are not supported with ARM instance types or classes, setting useOptimalInstanceClasses to false');
+    useOptimalInstanceClasses = false;
```

Comment:
are we changing the value here, i don't think we should modify it just throw a warning to let users know and prompt them to set it if its undefined

Reply:
My reasoning for modifying `useOptimalInstanceClasses` is because if we don't change the value, the error when deploying would occur. There's actually no case where useOptimalInstanceClasses can be True and there are arm instances like pahud mentioned `if instanceClasses is defined as all arm instance types, useOptimalInstanceClasses has to be false`. I believe it would make sense in this case to change it for the user since that's the only option. Let me know if you think it still makes more sense for just the warning!

Reply:
true, but we are kind of changing the field so if we want to do this i would say we might need to add feature flag, since we're kind of modifying the default behaviour now with setting it to `false` which i also agree is correct in this case, don;t want it to be intrusive for existing customers.

Reply:
Updated here: [e4b2864](https://github.com/aws/aws-cdk/pull/31510/commits/e4b286430863f31516705eaa22b1632ec3e8e5e9)!